system/include/compat/avx2intrin.h

```diff
@@ -0,0 +1,1800 @@
+    _mm256_set_m128d(                                                          \
+      (__m128d)wasm_i64x2_shuffle(                                             \
+        __a.v0, __a.v1, (((__imm) >> 4) & 3), (((__imm) >> 6) & 3)),           \
+      (__m128d)wasm_i64x2_shuffle(                                             \
+        __a.v0, __a.v1, ((__imm) & 3), (((__imm) >> 2) & 3)));                 \
+  })
+
+static __inline__ __m256
+  __attribute__((__always_inline__, __nodebug__, DIAGNOSE_SLOW))
+  _mm256_permutevar8x32_ps(__m256 __a, __m256i __b) {
+  __m256 ret;
+  int index[8];
+  float lane[8];
+  for (int i = 0; i < 4; i++) {
+    index[i] = ((__i32x4)__b.v0)[i] & 7;
+    index[i + 4] = ((__i32x4)__b.v1)[i] & 7;
+  }
+  for (int i = 0; i < 8; i++) {
+    lane[i] = index[i] < 4 ? ((__f32x4)(__a.v0))[index[i]]
+                           : ((__f32x4)(__a.v1))[index[i] - 4];
+  }
+  ret.v0 = (__m128)wasm_f32x4_make(lane[0], lane[1], lane[2], lane[3]);
+  ret.v1 = (__m128)wasm_f32x4_make(lane[4], lane[5], lane[6], lane[7]);
+  return ret;
+}
+
+#define _mm256_permute4x64_epi64(__A, __imm)                                   \
+  __extension__({                                                              \
+    __m256i __a = (__A);                                                       \
+    _mm256_set_m128i(                                                          \
+      wasm_i64x2_shuffle(                                                      \
+        __a.v0, __a.v1, (((__imm) >> 4) & 3), (((__imm) >> 6) & 3)),           \
+      wasm_i64x2_shuffle(                                                      \
+        __a.v0, __a.v1, ((__imm) & 3), (((__imm) >> 2) & 3)));                 \
+  })
+
+static __inline__ __m256i
+  __attribute__((__always_inline__, __nodebug__, DIAGNOSE_SLOW))
+  _mm256_permute2x128_si256(__m256i __a, __m256i __b, const int imm8) {
+  __m256i ret;
+  ret.v0 = __avx_select4i(__a, __b, imm8);
+  ret.v1 = __avx_select4i(__a, __b, imm8 >> 4);
+  return ret;
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm256_extracti128_si256(__m256i __a, const int imm8) {
+  if (imm8 & 0x1) {
+    return __a.v1;
+  } else {
+    return __a.v0;
+  }
+}
+
+static __inline__ __m256i __attribute__((__always_inline__, __nodebug__))
+_mm256_inserti128_si256(__m256i __a, __m128i __b, const int imm8) {
+  __m256i ret = __a;
+  if (imm8 & 0x1) {
+    ret.v1 = __b;
+  } else {
+    ret.v0 = __b;
+  }
+  return ret;
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_maskload_epi32(int32_t const* __p, __m128i __m) {
+  // This may cause an out-of-bounds memory load since we first load and
+  // then mask, but since there are no segmentation faults in Wasm memory
+  // accesses, that is ok (as long as we are within the heap bounds -
+  // a negligible limitation in practice)
+  // TODO, loadu or load, 128-bit align?
```

Comment:
Could this cause ASan failures?

Reply:
The snippet is ported from old code in avxintrin.h, I am not sure, I will check that.